tests/test_cluster.py

```diff
@@ -0,0 +1,628 @@
+        template = lambda x, wires: qml.RX(x, wires=0)
+
+        qc = qml.map(template, obs_list, dev)
+
+        assert len(qc) == 2
+
+        # evaluate cluster so that queue is populated
+        qc(1)
+
+        assert len(qc[0].ops) == 2
+        assert qc[0].ops[0].name == "RX"
+        assert qc[0].ops[1].name == "PauliX"
+
+        assert len(qc[1].ops) == 2
+        assert qc[1].ops[0].name == "RX"
+        assert qc[1].ops[1].name == "PauliY"
+
+        # test that device is broadcast
+        assert qc[0].device is qc[1].device
+
+    def test_mapping_over_devices(self):
+        """test that mapping over a list of devices produces
+        a QNodeCluster with the correct QNodes"""
+        dev_list = [qml.device("default.qubit", wires=1), qml.device("default.qubit", wires=1)]
+
+        obs_list = [qml.PauliX(0), qml.PauliY(0)]
+        template = lambda x, wires: qml.RX(x, wires=0)
+
+        qc = qml.map(template, obs_list, dev_list)
+
+        assert len(qc) == 2
+
+        # evaluate cluster so that queue is populated
+        qc(1)
+
+        assert len(qc[0].ops) == 2
+        assert qc[0].ops[0].name == "RX"
+        assert qc[0].ops[1].name == "PauliX"
+
+        assert len(qc[1].ops) == 2
+        assert qc[1].ops[0].name == "RX"
+        assert qc[1].ops[1].name == "PauliY"
+
+        # test that device is not broadcast
+        assert qc[0].device is not qc[1].device
+        assert qc[0].device is dev_list[0]
+        assert qc[1].device is dev_list[1]
+
+    def test_mapping_over_measurements(self):
+        """test that mapping over a list of measurement types produces
+        a QNodeCluster with the correct QNodes"""
+        dev = qml.device("default.qubit", wires=1)
+
+        obs_list = [qml.PauliX(0), qml.PauliY(0)]
+        template = lambda x, wires: qml.RX(x, wires=0)
+
+        qc = qml.map(template, obs_list, dev, measure=["expval", "var"])
+
+        assert len(qc) == 2
+
+        # evaluate cluster so that queue is populated
+        qc(1)
+
+        assert len(qc[0].ops) == 2
+        assert qc[0].ops[0].name == "RX"
+        assert qc[0].ops[1].name == "PauliX"
+        assert qc[0].ops[1].return_type == qml.operation.Expectation
+
+        assert len(qc[1].ops) == 2
+        assert qc[1].ops[0].name == "RX"
+        assert qc[1].ops[1].name == "PauliY"
+        assert qc[1].ops[1].return_type == qml.operation.Variance
+
+    def test_invalid_obserable(self):
+        """test that an invalid observable raises an exception"""
+        dev = qml.device("default.qubit", wires=1)
+
+        obs_list = [qml.PauliX(0), qml.S(wires=0)]
+        template = lambda x, wires: qml.RX(x, wires=0)
+
+        with pytest.raises(ValueError, match="Some or all observables are not valid"):
+            qml.map(template, obs_list, dev, measure=["expval", "var"])
+
+
+class TestApply:
+    """Tests for the apply function"""
+
+    @pytest.mark.parametrize("interface", ["autograd", "torch", "tf"])
```

Comment:
How come ``"numpy"`` is not in the list?

Reply:
Ah, because it activates the same logic as using `autograd` :) And because it's deprecated. So it was just me 'pre-optimizing' a bit. Do you think it's worth adding?

Reply:
I think so, until it is entirely deprecated, as currently, we expect it to work just fine.

Reply:
Okay, done